src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CSharp.Analyzers/BinaryFormatterUnsafeDeserializeAnalyzer.cs

```diff
@@ -0,0 +1,103 @@
+        private static readonly LocalizableString Description = new LocalizableResourceString(nameof(Resources.BinaryFormatterUnsafeDeserializeDescription), Resources.ResourceManager, typeof(Resources));
+
+        private static readonly DiagnosticDescriptor Rule = new(DiagnosticId, Title, MessageFormat, Category, DiagnosticSeverity.Warning, isEnabledByDefault: true, description: Description);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze);
+
+            context.RegisterSyntaxNodeAction(AnalyzeMemberAccessExpressions, SyntaxKind.SimpleMemberAccessExpression);
+        }
+
+        private void AnalyzeMemberAccessExpressions(SyntaxNodeAnalysisContext context)
+        {
+            var memberAccessExpression = (MemberAccessExpressionSyntax)context.Node;
+
+            // If the accessed member isn't named "UnsafeDeserialize" bail out
+            if (!TargetMember.Equals(memberAccessExpression.Name.ToString(), StringComparison.Ordinal))
+            {
+                return;
+            }
+
+            // Continue, only if the call is to a method called UnsafeDeserialize
+            if (!(memberAccessExpression.Parent is InvocationExpressionSyntax))
+            {
+                return;
+            }
+
+            // Get the identifier accessed
+            var accessedIdentifier = memberAccessExpression.Expression switch
+            {
+                IdentifierNameSyntax i => i,
+                ObjectCreationExpressionSyntax o => o.DescendantNodes().OfType<IdentifierNameSyntax>().FirstOrDefault(),
+                MemberAccessExpressionSyntax m => m.DescendantNodes().OfType<IdentifierNameSyntax>().LastOrDefault(),
+                _ => null
+            };
+
+            // Return if the accessed identifier wasn't from a simple member access expression, object creation expression, or identifier
+            if (accessedIdentifier is null)
+            {
+                return;
+            }
+
+            // If the accessed identifier resolves to a type symbol other than BinaryFormatter, then bail out
+            // since it means the user is calling some other similarly named API.
+            var accessedSymbol = context.SemanticModel.GetSymbolInfo(accessedIdentifier).Symbol;
+            if (accessedSymbol is ILocalSymbol localSymbol)
+            {
+                if (!TargetTypeSymbolName.Equals($"{localSymbol.Type.ContainingNamespace}.{localSymbol.Type.Name}", StringComparison.Ordinal))
```

Comment:
better to compare the namespace and type name separately - minimizes allocations in analyzers help with compilation performance (although here it may not be a big deal as at this point it has already filtered out a ton of nodes)